service/history/replication/ack_manager.go

```diff
@@ -279,27 +283,46 @@ func (p *ackMgrImpl) getTasks(
 		}
 
 		token = response.NextPageToken
+		skippedTasks := 0
 		for _, task := range response.Tasks {
-			if replicationTask, err := p.toReplicationTask(ctx, task); err != nil {
+			ns, err := p.namespaceRegistry.GetNamespaceByID(namespace.ID(task.GetNamespaceID()))
+			if err != nil {
+				if _, isNotFound := err.(*serviceerror.NamespaceNotFound); !isNotFound {
+					return nil, 0, err
+				}
+				// Namespace doesn't exist on this cluster (i.e. deleted). It is safe to skip the task.
+				skippedTasks++
+				continue
+			}
+			// If namespace doesn't exist on polling cluster, there is no reason to send the task.
+			if !ns.IsOnCluster(pollingCluster) {
+				skippedTasks++
+				continue
+			}
+
+			replicationTask, err := p.toReplicationTask(ctx, task)
+			if err != nil {
 				return nil, 0, err
-			} else if replicationTask != nil {
-				replicationTasks = append(replicationTasks, replicationTask)
+			} else if replicationTask == nil {
+				skippedTasks++
+				continue
 			}
+			replicationTasks = append(replicationTasks, replicationTask)
 		}
 
-		// break if seen at least one task or no more task
-		if len(token) == 0 || len(replicationTasks) > 0 {
+		// There are no more tasks in DB (token is empty) then we have what we have.
+		if len(token) == 0 {
 			break
 		}
-	}
-
-	// sanity check we will finish pagination or return some tasks
-	if len(token) != 0 && len(replicationTasks) == 0 {
-		p.logger.Fatal("replication task reader should finish pagination or return some tasks")
+		// If nothing was filtered out then we got all the tasks we need.
+		if skippedTasks == 0 {
+			break
+		}
+		batchSize = skippedTasks
```

Comment:
If there is one namespace that generate lots of replication tasks that needs to be filtered out, we could potentially run into issue that we would have to keep skipping and it may cause timeout which stop the caller side to make progress on this shard. That will cause backlog and trap the caller in this state forever. I think we need to have a max skip count to bail out.

Reply:
Put a max count to the outer for loop to avoid that problem.